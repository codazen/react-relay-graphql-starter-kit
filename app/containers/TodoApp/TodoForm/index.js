/* flow */

import React from 'react';

type Props = {
  addTodo: ?Function, // callback to create new todo in parent state
};

type State = {
  content: string, // content for new todo
};

/**
 * Form for creating a new todo item
 * Parent component: TodoApp
 * Child components: N/A
 */
export default class TodoForm extends React.Component {

  static defaultProps: Props;

  constructor(props: Props) {
    super(props);
    this.state = {
      content: '',
    };

    (this:any).handleChange = this.handleChange.bind(this);
    (this:any).handleSubmit = this.handleSubmit.bind(this);
  }

  state: State;

  props: Props;

  handleChange(event) {
    const newState = {};
    newState[event.target.name] = event.target.value;
    this.setState(newState);
  }

  handleSubmit(event) {
    const {
      addTodo,
    } = this.props;

    if (addTodo) {
      addTodo(this.state.content);
    }
    event.preventDefault();
    this.setState({
      content: '',
    });
  }

  render() {
    return (
      <div>
        <h2>Todo Form</h2>
        <input
          type='text'
          value={this.state.content}
          name='content'
          onChange={this.handleChange}
        />
        <button onClick={this.handleSubmit}>
          + Todo
        </button>
      </div>
    );
  }
}

TodoForm.defaultProps = {
  addTodo: () => {},
};
