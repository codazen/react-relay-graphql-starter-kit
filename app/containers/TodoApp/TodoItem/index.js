/* @flow */

import React from 'react';
import Relay from 'react-relay';
import ListItem from '../../../components/ListItem';

type Props = {
  todo: {
    content: string,
    id: string,
  },
};

class TodoItem extends React.Component {
  constructor(props: Props) { // eslint-disable-line
    super(props);
  }

  render() {
    return (
      <ListItem
        content={this.props.todo.content}
        id={this.props.todo.id}
      />
    );
  }
}

export default Relay.createContainer(TodoItem, {
  fragments: {
    todo: () => Relay.QL`
      fragment on Todo {
        id
        content
      }
    `,
  },
});
