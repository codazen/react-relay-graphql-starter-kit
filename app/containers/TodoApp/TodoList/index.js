/* @flow */

import React from 'react';
import Relay from 'react-relay';
import TodoItem from '../TodoItem';

type TodoEdge = {
  node: {
    id: string,
  },
};

type Props = {
  todos: {
    edges: Array<TodoEdge>,
  },
};

class TodoList extends React.Component {
  constructor(props: Props) { // eslint-disable-line
    super(props);
  }

  render() {
    return (
      <div>
        <h2>Todo List</h2>
        {this.props.todos.edges.map(edge =>
          <TodoItem key={edge.node.id} todo={edge.node} />,
        )}
      </div>
    );
  }
}

export default Relay.createContainer(TodoList, {
  fragments: {
    todos: () => Relay.QL`
      fragment on TodoConnection {
        edges {
          node {
            id
            ${TodoItem.getFragment('todo')}
          }
        }
      }
    `,
  },
});
