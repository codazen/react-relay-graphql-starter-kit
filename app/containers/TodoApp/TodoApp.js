/* @flow */

import React from 'react';
import Relay from 'react-relay';
import TodoForm from './components/TodoForm';
import TodoList from './components/TodoList';
import AddTodoMutation from './mutations/AddTodoMutation';

type Props = {
  user: {
    todos: any,
  },
  relay: {
    commitUpdate: Function, // function to make mutation call to relay store
  },
};

/**
 * Container for todo application
 * Parent component: N/A
 * Child components: TodoList, TodoForm
 */
class TodoApp extends React.Component {

  props: Props;

  /**
   * addTodo function passed to the TodoForm
   * @param {content: string} content for new todo
   */
  handleAddTodo = (content: string) => {
    this.props.relay.commitUpdate(
      new AddTodoMutation({ content, user: this.props.user }),
    );
  }

  render() {
    return (
      <div className="todo-app-component">
        <h1>TODO APP</h1>
        <TodoForm addTodo={this.handleAddTodo} />
        <TodoList todos={this.props.user.todos} user={this.props.user} />
      </div>
    );
  }
}

exports.TodoApp = TodoApp;

export default Relay.createContainer(TodoApp, {
  fragments: {
    user: () => Relay.QL`
      fragment on User {
        id
        todos(first: 1000) {
          ${TodoList.getFragment('todos')}
        }
        ${TodoList.getFragment('user')}
        ${AddTodoMutation.getFragment('user')}
      }
    `,
  },
});
