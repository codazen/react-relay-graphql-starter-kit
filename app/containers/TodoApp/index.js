/* @flow */

import React from 'react';
import Relay from 'react-relay';
import TodoForm from './TodoForm';
import TodoList from './TodoList';

type Props = {
  user: {
    todos: any,
  },
};

/**
 * Container for todo application
 * Parent component: N/A
 * Child components: List, TodoForm
 */
class TodoApp extends React.Component {

  constructor(props: Props) { // eslint-disable-line
    super(props);
  }

  props: Props;

  render() {
    return (
      <div>
        <h1>Todo App</h1>
        <TodoForm />
        <TodoList todos={this.props.user.todos} />
      </div>
    );
  }
}

export default Relay.createContainer(TodoApp, {
  fragments: {
    user: () => Relay.QL`
      fragment on User {
        todos(first: 1000) {
          ${TodoList.getFragment('todos')}
        }
      }
    `,
  },
});
