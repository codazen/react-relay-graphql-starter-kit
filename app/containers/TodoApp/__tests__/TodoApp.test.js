import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { makeExecutableSchema, addMockFunctionsToSchema, MockList } from 'graphql-tools';
import { graphql } from 'graphql';
import casual from 'casual';
import schemaString from 'data/schema.graphql';
import { TodoApp } from '../TodoApp';
import AddTodoMutation from '../mutations/AddTodoMutation';

// setup React as global variable
global.React = React;

// use consistent seed for snapshot matching
casual.seed(123);

// create schema based on server/data/schema.graphql
const schema = makeExecutableSchema({ typeDefs: schemaString });

// mock data
const numListItems = 10;
const mocks = {
  User: () => ({
    id: () => casual.integer(0, 1000).toString(),
  }),
  TodoConnection: () => ({
    edges: () => new MockList(numListItems),
  }),
  Todo: () => ({
    id: () => casual.integer(0, 1000).toString(),
    content: () => casual.string,
  }),
};

// add mocks to schema in place
addMockFunctionsToSchema({
  schema,
  mocks,
});

describe('TodoApp component', () => {
  let todoApp;
  let user;
  const query = `query
    TodoApp {
      user {
        id
        todos(first: 10) {
          edges {
            node {
              id
              content
            }
          }
        }
      }
    }
  `;

  // mock mutation calls and use _unresolvedProps to detect failure
  const AddTodoStatus = jest.fn(mutation => mutation._unresolvedProps.content); // eslint-disable-line
  const mockSuccessFn = jest.fn((onSuccess, transaction) => onSuccess(transaction));
  const mockFailureFn = jest.fn((onFailure, transaction) => onFailure(transaction));

  const relay = {
    commitUpdate: jest.fn((mutation, callbacks) => {
      if (mutation instanceof AddTodoMutation) {
        return AddTodoStatus(mutation)
          ? mockSuccessFn(callbacks ? callbacks.onSuccess : () => {}, {})
          : mockFailureFn(callbacks ? callbacks.onFailure : () => {}, {});
      }
      return {};
    }),
  };

  beforeEach(async () => {
    await graphql(schema, query).then((result) => {
      user = result.data.user;
      todoApp = shallow(<TodoApp user={user} relay={relay} />);
      AddTodoStatus.mockClear();
      mockSuccessFn.mockClear();
      mockFailureFn.mockClear();
    });
  });

  it('Renders correctly', () => {
    expect(shallowToJson(todoApp)).toMatchSnapshot();
  });

  describe('Unit test: handleAddTodo', () => {
    it('Success', () => {
      const content = 'testContent';
      todoApp.instance().handleAddTodo(content);
      expect(todoApp.instance().props.relay.commitUpdate).toHaveBeenCalled();
      expect(AddTodoStatus).toHaveBeenCalled();
      expect(mockSuccessFn).toHaveBeenCalled();
    });

    it('Fails', () => {
      const content = undefined;
      todoApp.instance().handleAddTodo(content);
      expect(todoApp.instance().props.relay.commitUpdate).toHaveBeenCalled();
      expect(AddTodoStatus).toHaveBeenCalled();
      expect(mockFailureFn).toHaveBeenCalled();
    });
  });
});
