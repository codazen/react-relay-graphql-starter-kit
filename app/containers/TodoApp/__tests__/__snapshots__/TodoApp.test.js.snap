exports[`TodoApp component Renders correctly 1`] = `
<div
  className="todo-app-component">
  <h1>
    TODO APP
  </h1>
  <TodoForm
    addTodo={[Function]} />
  <Relay(TodoList)
    todos={
      Object {
        "edges": Array [
          Object {
            "node": Object {
              "content": "quis molestiae tempora est omnis porro porro",
              "id": "713",
            },
          },
          Object {
            "node": Object {
              "content": "nihil omnis reiciendis velit nobis culpa deleniti",
              "id": "491",
            },
          },
          Object {
            "node": Object {
              "content": "reprehenderit in nisi et qui quia odio",
              "id": "140",
            },
          },
          Object {
            "node": Object {
              "content": "veritatis eligendi in deserunt id harum sit",
              "id": "245",
            },
          },
          Object {
            "node": Object {
              "content": "dolor dicta provident quo provident molestiae harum",
              "id": "440",
            },
          },
          Object {
            "node": Object {
              "content": "pariatur quos est quod laborum ea quisquam",
              "id": "298",
            },
          },
          Object {
            "node": Object {
              "content": "quia et commodi autem incidunt ut exercitationem",
              "id": "417",
            },
          },
          Object {
            "node": Object {
              "content": "et voluptas explicabo omnis iusto doloremque et",
              "id": "988",
            },
          },
          Object {
            "node": Object {
              "content": "quos sequi molestiae beatae enim necessitatibus molestiae",
              "id": "174",
            },
          },
          Object {
            "node": Object {
              "content": "saepe eligendi earum excepturi et laudantium fuga",
              "id": "750",
            },
          },
        ],
      }
    }
    user={
      Object {
        "id": "696",
        "todos": Object {
          "edges": Array [
            Object {
              "node": Object {
                "content": "quis molestiae tempora est omnis porro porro",
                "id": "713",
              },
            },
            Object {
              "node": Object {
                "content": "nihil omnis reiciendis velit nobis culpa deleniti",
                "id": "491",
              },
            },
            Object {
              "node": Object {
                "content": "reprehenderit in nisi et qui quia odio",
                "id": "140",
              },
            },
            Object {
              "node": Object {
                "content": "veritatis eligendi in deserunt id harum sit",
                "id": "245",
              },
            },
            Object {
              "node": Object {
                "content": "dolor dicta provident quo provident molestiae harum",
                "id": "440",
              },
            },
            Object {
              "node": Object {
                "content": "pariatur quos est quod laborum ea quisquam",
                "id": "298",
              },
            },
            Object {
              "node": Object {
                "content": "quia et commodi autem incidunt ut exercitationem",
                "id": "417",
              },
            },
            Object {
              "node": Object {
                "content": "et voluptas explicabo omnis iusto doloremque et",
                "id": "988",
              },
            },
            Object {
              "node": Object {
                "content": "quos sequi molestiae beatae enim necessitatibus molestiae",
                "id": "174",
              },
            },
            Object {
              "node": Object {
                "content": "saepe eligendi earum excepturi et laudantium fuga",
                "id": "750",
              },
            },
          ],
        },
      }
    } />
</div>
`;
